Update WASM binary and Land WASM rust sources in this repository

ochameau opened the pull request:
Today, the rust sources used to [generate lib/mappings.wasm](https://github.com/mozilla/source-map/blob/master/lib/mappings.wasm) are hosted on:
https://github.com/fitzgen/source-map-mappings/
This looks unofficial, whereas that's where the sources come from and last binary was built from that repo.

It would be easier to fold the rust sources in this repo to ease contributions to it, while increasing visibility to the rust sources.

While we are at landing sources in-tree, it would also be nice to refresh the wasm binary by building it with latest rust version.
The current binary has been being in 2018 and is no longer reproducible.

In an effort to help reproduce the same wasm file, here is the environment from which I built it:
```
$ cargo -V
cargo 1.64.0 (387270bc7 2022-09-16)
$ rustup -V
rustup 1.25.1 (bb60b1e89 2022-07-12)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.64.0 (a55dd71d5 2022-09-19)`
```
Using instruction documented [here](https://github.com/mozilla/source-map/blob/master/CONTRIBUTING.md#updating-the-libmappingswasm-webassembly-module).

ochameau commented on wasm-mappings/.gitattributes:
```diff
@@ -0,0 +1,2 @@
+README.md -diff -merge
```
Oh yes I forgot to remove it after folding it into the main one.

ochameau commented on wasm-mappings/LICENSE-APACHE:
```diff
@@ -0,0 +1,201 @@
+                              Apache License
```
I'll reach out before landing. I suspect they changed this repo license without changing this other repo.

fitzgen replied:
I am happy to relicense the rust code as BSD.

ochameau replied:
Thanks a lot for chiming in Nick!

ochameau commented:
> RSLGTM on the Rust and Python code (assuming that it's just copying existing work).

Yes I copied master branch as is.

ochameau commented:
I removed the license files from wasm folder, updated various files to mention BSD license, and updated github repo reference to mozilla's source-map one.

bomsy commented on wasm-mappings/CONTRIBUTING.md:
```diff
@@ -0,0 +1,87 @@
+# Contributing to `source-map-mappings`
+
+Hi! We'd love to have your contributions! If you want help or mentorship, reach
+out to us in a GitHub issue, or ping `fitzgen` in [#rust on irc.mozilla.org](irc://irc.mozilla.org#rust)
```
Is this irc channel still valid?

ochameau replied:
irc server is down, I removed it with the new matrix channel.

bomsy commented on wasm-mappings/README.md:
```diff
@@ -0,0 +1,33 @@
+# `source-map-mappings`
```
Sould this become `wasm-mappings`?

ochameau replied:
No, the rust package name is still source-map-mappings (you can see it referenced in cargo files)

bomsy reviewed (approved):
LGTM! Just had comments/questions about documentation updates
Thanks

ochameau commented on CONTRIBUTING.md:
````diff
@@ -152,11 +152,11 @@ $ rustup toolchain install nightly
+Move to wasm sources folder:
 
 ```
 $ git clone https://github.com/fitzgen/source-map-mappings.git
````
yes, forgot to remove the command line!

ochameau commented:
I've also updated the patch to ensure using latest master revision of source-map-mappings.
So that we are using https://github.com/fitzgen/source-map-mappings/commit/197660faf3dd1167c643797a5ed62cb687482c54 here, with the few cleanups and renames.

ochameau commented:
With the latest changeset refreshing the wasm binary with latest version of rust, there seem to be a significant performance improvement:
https://treeherder.mozilla.org/perfherder/comparesubtest?originalProject=try&newProject=try&newRevision=340b70fc0526e767e8657b8f41749240d14d2908&originalSignature=4081483&newSignature=4081483&framework=12&originalRevision=b4fda98f266ca9b53c8f3962d6e92a9d68e32b57&page=1&showOnlyImportant=1
![image](https://user-images.githubusercontent.com/199907/197014367-207a3684-2506-4fa4-b472-24b0ad14d428.png)
(I ported the existing perf test from bench folder to firefox CI)

eemeli reviewed (approved):
Looks good, provided that the contribution guidelines and the README link to them is fixed.

ochameau commented:
When testing against Babel and React sourcemaps there is also significant improvements when running on Firefox:
![image](https://user-images.githubusercontent.com/199907/197772307-03a616f5-810a-46ad-b991-f120d541cbb0.png)
Up to 20% faster when parsing the babel sourcemap and between 5% to 7% faster on mapping generated to original locations.

ochameau commented on wasm-mappings/README.tpl:
```diff
@@ -0,0 +1,3 @@
+# `{{crate}}`
+
+{{readme}}
```
I imagine there must be benefits to use this setup when more than one file contribute to the README.md.
But in the current state of things, I agree it sounds easier.

ochameau commented on wasm-mappings/CONTRIBUTING.md:
```diff
@@ -0,0 +1,87 @@
+# Contributing to `source-map-mappings`
```
I dropped the redundant content, but kept it for wasm specifics instructions.
I may consider moving "updating mappings.wasm" here?
I find it nice to keep the rust/wasm part of the codebase still contained in this subfolder.

eemeli replied:
Yeah, that makes sense.

ochameau commented:
I had to request the enabling of `actions-rs/toolchain@v1` on mozilla repos for the wasm github actions:
https://bugzilla.mozilla.org/show_bug.cgi?id=1797457
Otherwise the github actions for wasm were failing on rust tests.

Now everything is green and good to go!
